### ISSUE TITLE
Extended Tweets: Entities appear incorrect when received over Streaming API

### ISSUE THREAD

**Issue by JoshKeegan** [linvi/tweetinvi#344]
Hi Linvi,
When I receive an extended Tweet over the Streaming API, it has the new `ExtendedTweet` property.
Main question: should I be using this myself, or should Tweetinvi be putting this data into the appropriate fields in the returned implementation of `ITweet` (if in extended tweets mode)? 

Second question: In the `ExtendedTweet` property there are `LegacyEntities` and `ExtendedEntities`. For the example i'm looking at `ExtendedEntities` are null, and `LegacyEntities` appear to contain the correct set of entities. Using https://github.com/twitterdev/tweet-updates/blob/master/samples/original/%3E140%20characters%20(hidden%20%2B%20extended)/attachment.json as a sample Tweet JSON, I assume `LegacyEntities` is the `entities` property of `extended_tweet`, but what is `ExtendedEntities` to be used for? I'm assuming there is some use, but I just haven't found it :)
Cheers,
Josh

**Comment by linvi** [comment]
> When I receive an extended Tweet over the Streaming API, it has the new ExtendedTweet property.
Main question: should I be using this myself, or should Tweetinvi be putting this data into the appropriate fields in the returned implementation of ITweet (if in extended tweets mode)?

Tweetinvi will handling it for you. If a Tweet coming from a stream contains the `extended_tweet`, Tweetinvi will put the data directly into the main `Tweet`.

> Second question: In the ExtendedTweet property there are LegacyEntities and ExtendedEntities. For the example i'm looking at ExtendedEntities are null, and LegacyEntities appear to contain the correct set of entities. Using https://github.com/twitterdev/tweet-updates/blob/master/samples/original/%3E140%20characters%20(hidden%20%2B%20extended)/attachment.json as a sample Tweet JSON, I assume LegacyEntities is the entities property of extended_tweet, but what is ExtendedEntities to be used for? I'm assuming there is some use, but I just haven't found it :)

All the implementation is based on the documentation (https://dev.twitter.com/overview/api/upcoming-changes-to-tweets > Compatibility Mode JSON Rendering). 

Twitter team says that Extended Tweets have not been fully implemented yet. What Tweetinvi does at the current time is that the `Tweet.Entities` will contain a merge of the following :

* `Tweet.LegacyEntities`
* `Tweet.ExtendedTweet.LegacyEntities`
* `Tweet.ExtendedTweet.ExtendedEntities`

In version 1.2, I will look at how they are structured in real life (as opposed to documentation) and do any improvements that I can.

**Comment by JoshKeegan** [comment]
Great :) Having Tweetinvi push these into the main object will make things simpler (for me).

Can you take a look at the Entities.Urls property? 
It currently contains a link to the status, which I think should only be in the truncated version of the tweet, not the extended one.
This is what got me confused about this :)

e.g. 
![0 tweetinvi logic twitterentities urlentity](https://cloud.githubusercontent.com/assets/1682204/18991234/10b07ffa-870f-11e6-8b96-276039360aec.jpg)

**Comment by linvi** [comment]
This has been done intentionally.
Maybe I will change this later in but my idea was that developers do not need to look at the legacy or extended entities.

They are all combined in one place `Tweet.Entities`.

**Comment by JoshKeegan** [comment]
Right, this creates a bit of a problem for me because in some of my code, when a status gets displayed to the user the entities will be replaced with equivalent links, pictures etc...

In this case, this URL is an entity that doesn't exist in the status full text.
This means that when I replace the text between the specified entity indices with a HTML anchor, it will actually be replacing random text in the status.
In addition to that, there's a more subtle problem: entity indices can overlap when combined from two source strings (as you're doing), which is impossible in the same string. This means my replacement code wouldn't even run for this example tweet, as it would be trying to replace the same region of the string twice.

I'd expect `Tweet.Entities` to be the extended entities (and only extended ones) when Tweetinvi is in extended tweet mode. 
My understanding is that anyone needing the entities for the truncated text wouldn't be using extended tweets. Is there some use case for them that I've missed?

**Comment by linvi** [comment]
I think you have a good point here and I think I do agree with you. 

Can you please use the `ExtendedTweet` for the moment as I would prefer to release this as part of 1.2 after having maybe a bit more feedback on this.

**Comment by JoshKeegan** [comment]
Alright, thanks :)

**Comment by JoshKeegan** [comment]
Hi Linvi!
I've come back to this because after I first opened the issue I just stopped using the Streaming API, with everything falling back to REST API polling, so I hadn't tested any changes since then.
I've started trying to get a working build using the Streaming API again this afternoon by using Tweetinvi 1.2.0.1, but came accross this same issue again.

I'm going to open a PR with my changes, which also contains a couple of other minor bits to (hopefully) make it easier to use Tweetinvi, Extended Tweets and the Streaming API together.
Cheers,
Josh
